graph-builder/src/registry.rs

```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(
+    credentials_file: &str,
+) -> Result<Box<dyn std::io::Read>, Error> {
+    if credentials_file != "" {
+        Ok(Box::new(File::open(credentials_file)?))
+    } else {
+        Ok(Box::new(std::io::empty()))
+    }
+}
+
+pub fn authenticate_client<'a>(
+    client: &'a mut dkregistry::v2::Client,
+    login_scope: &'a str,
+) -> impl futures::future::Future<Item = &'a dkregistry::v2::Client, Error = dkregistry::errors::Error>
+{
+    futures::future::ok(client)
```

Comment:
This isn't needed. You can start the future chain with the call to `dclient.is_v2_supported()` below.